Store syntax state on every line

bynect opened the pull request:
Store a struct SHSTATE on every line (only 8 additional bytes on my machine) so that syntaxHighlight doesn't need to loop on the whole file for every modification, but only from cy - 1 to num_lines. (*)

(*) This  can be further improved by stopping the highlighting when no multiline comment/string/etc is modified and returning early.

~~Seems to work fine for what I have tested.~~
This is also related to #59

bynect commented:
Just found a bug, I'll try to fix it

ArthurBacci64 commented:
Fixed?

ArthurBacci64 commented:
Looks like matching brackets are not working.

bynect commented:
> Looks like matching brackets are not working.

I should have fixed it. In which circumstances the matching brackets are not working? They seem to be working properly for me

ArthurBacci64 commented:
> > Looks like matching brackets are not working.
> 
> I should have fixed it. In which circumstances are the matching brackets not working? They seem to be working properly for me

I just opened a file with the branch `optimize2` of your fork. I did `git pull` and recompiled, still not working.

bynect commented:
> > > Looks like matching brackets are not working.
> > 
> > 
> > I should have fixed it. In which circumstances are the matching brackets not working? They seem to be working properly for me
> 
> I just opened a file with the branch `optimize2` of your fork. I did `git pull` and recompiled, still not working.

But what is the problem? I just tried on a .cpp file and it seems like brackets are matched and highlighted correctly.

ArthurBacci64 commented:
> > > > Looks like matching brackets are not working.
> > > 
> > > 
> > > I should have fixed it. In which circumstances are the matching brackets not working? They seem to be working properly for me
> > 
> > 
> > I just opened a file with the branch `optimize2` of your fork. I did `git pull` and recompiled, still not working.
> 
> Yes, but what is the problem? I just tried on a .cpp file and it seems like brackets are matched and highlighted correctly.

The problem is that the brackets are not being highlighted in my machine.

bynect commented:
> > > > > Looks like matching brackets are not working.
> > > > 
> > > > 
> > > > I should have fixed it. In which circumstances are the matching brackets not working? They seem to be working properly for me
> > > 
> > > 
> > > I just opened a file with the branch `optimize2` of your fork. I did `git pull` and recompiled, still not working.
> > 
> > 
> > Yes, but what is the problem? I just tried on a .cpp file and it seems like brackets are matched and highlighted correctly.
> 
> The problem is that the brackets are not being highlighted in my machine.

![brackx](https://user-images.githubusercontent.com/68197565/113784077-10738880-9735-11eb-99e8-54300cf7584e.PNG)
![xxx2](https://user-images.githubusercontent.com/68197565/113784080-123d4c00-9735-11eb-900d-68234d7b4f6c.PNG)
I see them like this (with c syntax)

ArthurBacci64 commented:
Seem only to update when a change is made in the buffer. It should update the matching brackets when the cursor changes it position.

bynect commented:
> Seem only to update when a change is made in the buffer. It should update the matching brackets when the cursor changes it position.

Oh I see the problem now. This is because of the way state is passed from a line to the next without iterating all of them. I guess I can fix this problem by backtracking to the line of the matching bracket and updating its color

bynect commented:
Also I was planning to fix this soon
> (*) This can be further improved by stopping the highlighting when no multiline comment/string/etc is modified and returning early.

bynect commented:
> Also I was planning to fix this soon
> 
> > (*) This can be further improved by stopping the highlighting when no multiline comment/string/etc is modified and returning early.

This should be done and seems to work fine, aka updating only the line of the changes if no comment/string are present

bynect commented:
I'm not sure if the problem with brackets was solved in ac8c6987c652b1f33286cef62dc60a39a8289629, could you test it?

ArthurBacci64 commented:
Still not working, by the way there's an overflow.

ArthurBacci64 commented:
When I changed a part of the file, asan detected this overflow.
```
=================================================================
==5036==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x627000065410 at pc 0x55be04581992 bp 0x7fffc7144b30 sp 0x7fffc7144b20
READ of size 8 at 0x627000065410 thread T0
    #0 0x55be04581991 in syntaxHighlight (/home/arthur/repos/Teditor/ted+0x1e991)
    #1 0x55be0457dc99 in main (/home/arthur/repos/Teditor/ted+0x1ac99)
    #2 0x7f816b1dbb24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
    #3 0x55be0457e11d in _start (/home/arthur/repos/Teditor/ted+0x1b11d)

0x627000065410 is located 32 bytes to the right of 13040-byte region [0x627000062100,0x6270000653f0)
allocated by thread T0 here:
    #0 0x7f816b4a483a in __interceptor_realloc /build/gcc/src/gcc/libsanitizer/asan/asan_malloc_linux.cpp:164
    #1 0x55be04588481 in read_lines (/home/arthur/repos/Teditor/ted+0x25481)

SUMMARY: AddressSanitizer: heap-buffer-overflow (/home/arthur/repos/Teditor/ted+0x1e991) in syntaxHighlight
Shadow bytes around the buggy address:
  0x0c4e80004a30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4e80004a40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4e80004a50: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4e80004a60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4e80004a70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fa fa
=>0x0c4e80004a80: fa fa[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4e80004a90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4e80004aa0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4e80004ab0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4e80004ac0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4e80004ad0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==5036==ABORTING
```

bynect commented:
Overflow fixed, I'll into a way to fix brackets

bynect commented:
Should be fixed with 3c907a1. I added a way to store the positions of the highlighted brackets that will be reset when the cursor is moved, though this is not the best approach for solving this problem, it's the best I could come up with.

bynect commented:
Upon closer analysis the previous commit didn't fixed the problem and added complexity, so I reverted it.
For now I have added a "temporary" solution, basically rerun the syntax highlighting when the cursor is moved but only if the character on which the cursor is pointing on is one of the character of `current_syntax.match` (previously syntaxHighlight was reset to zero on every cursor position change)

ArthurBacci64 commented:
Brackets that come after numbers are coloured.
![bugteditor](https://user-images.githubusercontent.com/59983362/114394567-ba0dac00-9b71-11eb-95ee-b8ccbf1fd2e4.png)

bynect commented:
I don't see a clear solution to this problem right now, and for the sake of simplicity, I am going to close (at least for now). I made another pr with a commit  that didn't break parenthesis
